backend/dataall/modules/shares_base/services/sharing_service.py

```diff
@@ -0,0 +1,327 @@
+import logging
+from typing import List, Union
+from abc import ABC, abstractmethod
+from sqlalchemy import and_
+from time import sleep
+from dataall.base.db import Engine
+from dataall.base.aws.iam import IAM
+from dataall.core.environment.services.environment_service import EnvironmentService
+from dataall.modules.shares_base.db.share_object_state_machines import (
+    ShareObjectSM,
+    ShareItemSM,
+)
+from dataall.modules.shares_base.services.shares_enums import (
+    ShareItemHealthStatus,
+    ShareObjectActions,
+    ShareItemStatus,
+    PrincipalType,
+)
+from dataall.modules.shares_base.db.share_object_models import ShareObjectItem, ShareObject
+from dataall.modules.shares_base.db.share_object_repositories import ShareObjectRepository
+from dataall.modules.datasets_base.db.dataset_models import DatasetLock
+
+log = logging.getLogger(__name__)
+
+MAX_RETRIES = 10
+RETRY_INTERVAL = 60
+
+
+class SharesManagerInterface(ABC):
+    @abstractmethod
+    def __init__(
+        self,
+        session,
+        dataset,
+        share,
+        items,
+        source_environment,
+        target_environment,
+        source_env_group,
+        env_group,
+        reapply,
+    ):
+        """Ensures that the manager used is initalized with the paramters passed"""
+        ...
+
+
+class SharesProcessorInterface(ABC):
+    @staticmethod
+    @abstractmethod
+    def initialize_share_manager(
+        session, dataset, share, items, source_environment, target_environment, source_env_group, env_group, reapply
+    ) -> Union[SharesManagerInterface, List[SharesManagerInterface]]:
```

Comment:
I think using only `List[SharesManagerInterface]` is cleaner and will simplify the implementations. If caller wants to pass one instance they will have to wrap it in a list